### PR TITLE
support krb5 1.20

### PR DESCRIPTION
**Pull request by abbra**
Add a basic support for krb5 1.20 version. 1.20 has changes to KDB API, aimed at a better way of issuing PACs in the tickets. As a result, `sign_authdata()` callback was replaced with an `issue_pac()` callback. In this patchset we handle both pre-1.20 and 1.20 versions of MIT Kerberos with a single exception: RBCD support is not implemented yet. It is a new optional API callback for KDB API of krb5 1.20 and it will come later.

Since Fedora does not yet have krb5 1.20, a separate COPR repo is provided with a test rebuild of krb5 and samba: `abbra/krb5-test`. This pull request builds against stock Fedora krb5 version (1.19).

Following PRs in the parallel CI infrastructure are used to test both 1.19 and 1.20 versions:
- https://github.com/freeipa-pr-ci2/freeipa/pull/1776 uses `abbra/krb5-test` COPR repo to run against krb5 1.20 on Fedora 37
- https://github.com/freeipa-pr-ci2/freeipa/pull/1770 uses stock Fedora 37 krb5 1.19 version.

Both cases fail with S4U operations against Active Directory due to a separate issue. This is expected and cannot be solved in FreeIPA: a KDC checksum in PAC buffer is derived using a most recent encryption key that KDC has access to. In FreeIPA environment we default to encryption types from RFC8009 (SHA-2-based) while Active Directory does not support them at all and requires the KDC checksum use AES256-SHA-1 at most. This issue is tracked by a separate MIT Kerberos issue upstream.

### CODE REVIEW

**Comment by abbra** [comment]
So far, 151 test suite pased, 8 failed for non-related reasons. ACME test suite timed out.

**Comment by abbra** [comment]
A run at the parallel infrastructure for krb5 1.20 COPR also produced 2 errored, 3 failing, 8 pending, and 135 successful test suites. I am re-running failed/errored tests as it looks like an infrastructure is a bit at a strain with multiple parallel runs. In general, there are no tests that failed due to this change.

**Comment by abbra** [comment]
@stanislavlevin @rcritten @sumit-bose  @jrisc could you please review this PR?

**Comment by abbra** [comment]
At least 20100 is similar to https://pagure.io/freeipa/issue/9251 where LDAP context gets NULL-ed in a parallel request of something else from LDAP. 

28371 looks like a shutdown of verto event loop processing went wrong in ev backend.

**Comment by abbra** [comment]
Yes, 20100 is exactly issue 9251, after calling into `is_master_host()` we get `ipactx->lcontext` NULL-ed and since it is not checked, we pass NULL to `ipadb_ldap_deref_results()`. It is unrelated to krb5 1.20 port at all -- we had these disappearing LDAP context issues before and were patching over with checks for the LDAP context but never got to the real reason why it fails.

The issue 9251 is one of closest to the real cause. `is_master_host()` runs `ipadb_simple_search()` which might get to `ipadb_need_retry()` and that one might get get to `ipadb_get_connection()` which unconditionally frees existing connection. The problem might be in the logic in `ipadb_simple_*()` functions where in case of a failure to perform the operation we may leave LDAP context removed (NULL-ed) and the consumer needs to handle it. If that happened when processing some other LDAP request, we'd fail completely because there is no original LDAP context anymore.

File: daemons/ipa-kdb/ipa_kdb_delegation.c [hunk 1]
```diff
@@ -106,20 +106,26 @@ static krb5_error_code ipadb_match_acl(krb5_context kcontext,
...
 
-    kerr = krb5_unparse_name(kcontext, client, &client_princ);
-    if (kerr != 0) {
-        goto done;
-    }
-    kerr = krb5_unparse_name(kcontext, target, &target_princ);
-    if (kerr != 0) {
-        goto done;
+    if ((client != NULL) && (target != NULL)) {
+        kerr = krb5_unparse_name(kcontext, client, &client_princ);
```
Comment (jrisc) [inline daemons/ipa-kdb/ipa_kdb_delegation.c:119]: `kerr` is not initialized but set unconditionally in the middle of the function. I think if would be safer to define it with an initial value of `ENOENT` and remove this assignation in the middle of the function.

Given the fact `kerr` is return systematically at the end of the function, we should know what is its default value by looking at its definition without having to go through the entire function.

Comment (jrisc) [reply daemons/ipa-kdb/ipa_kdb_delegation.c:119]: I withdraw this one. The `kerr` assignment in the middle of the function is actually necessary in order to reset the *failing by default* state after success exit code assignments from `krb5_unparse_name()` calls.

File: daemons/ipa-kdb/ipa_kdb_mspac.c [hunk 2]
```diff
@@ -2250,42 +2242,15 @@ krb5_error_code ipadb_verify_pac(krb5_context context,
...
-    kerr = krb5_pac_verify_ext(context, old_pac, authtime,
-                               client_princ, srv_key, priv_key,
-                               (is_cross_realm &&
-                                (flags & KRB5_KDB_FLAG_PROTOCOL_TRANSITION)));
-    if (kerr) {
-        goto done;
     }
 
-    /* Now that the PAC is verified, do additional checks.
+    /* In krb5 1.20 the PAC signatures are verified prior to call to issue_pac().
```
Comment (rcritten) [inline daemons/ipa-kdb/ipa_kdb_mspac.c:2307]: nit: did you want to specify krb5 1.20+ as you've done elsewhere in the code?

**Comment by rcritten** [comment]
Thanks, these changes LGTM.

**Comment by abbra** [comment]
master:

* 5e7590981b60c266ba87a2427633f14b3a3b6124 ipa-kdb: refactor MS-PAC processing to prepare for krb5 1.20
* c1582bd322d239e5270974da5a3678dbba458599 ipa-kdb: add krb5 1.20 support
* a9018da90d94585a437a2d1730454c5afc36b39f ipa-kdb: handle cross-realm TGT entries when generating PAC
* e86807b58c67a607bceb568d206af3b3abc03dea ipa-kdb: handle empty S4U proxy in allowed_to_delegate
* 0c67f0e60766864ec9416a18716e90ff46b5020e ipa-kdb: fix PAC requester check
* ce05e5fd40e63262e115d7adc4a21915dc2bb32d ipa-kdb: fix comment to make sure we talk about krb5 1.20 or later

**Pull request closed without merging**